### PR TITLE
Generate reports about profiling results of top-level sections of the application.

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

See: [#180311774](https://www.pivotaltracker.com/story/show/180311774).

Usage:
- Capturing profiling data: Ctrl+Alt+p at any time in the IDE will dump a JSON eventlog to the console. Detail level is determined by an environment variable at compile time: `ENSO_MAX_PROFILING_LEVEL=debug` will enable maximum detail; by default only Objective-level profilers are enabled.
- Rendering a graph:  
  - A `render-profile` debug scene uses the `flame_graph` libraries to render a log. For now, the eventlog is read from a hard-coded path at `dist/content/profile.json` in the enso root directory.
- Generating a text report:
  - A [command-line Measurement pretty-printer](https://github.com/enso-org/enso/blob/wip/kw/profiling-application/lib/rust/profiler/data/src/bin/measurements.rs) illustrates the hierarchy of profilers. It reads a JSON eventlog from stdin. ([Example output below](https://github.com/enso-org/enso/pull/3306#issuecomment-1058533022))

### Important Notes

Instrumenting:
- Added enough instrumentation to get a sense of where we're spending our time when starting up and creating or opening a project.
- This PR only uses the high-level API; we'll need to use the low-level API to connect the notification-sender to the notification handler(s) for e.g. ProjectOpened. I'm deferring that because the low-level API could use some changes that don't need to be in the scope of this PR.
- This PR refactors some `async` blocks to `async fn`s so that #[profile] can be applied to them. I think this is usually a reasonable refactoring if the block is not trivial. If the async block is trivial, it would be nice to be able to profile it without promoting it to a fn. #[profile] could support this by automatically wrapping async blocks within a profiled fn. That's a topic for another PR.

Design concepts:
- I'm using `Objective` profilers a bit differently from the design: I think we should consider a user-visible delay to correspond to the interval from the beginning of some Objective A, to the finish of some Objective B, where B can be A or another profiler that should be a descendant of A. This will allow some simplification to the low-level API, and also allow us to use the high-level API in many places where we'd need the low-level API in the current design.

Implementation:
- Fixed some unhandled cases in #[profile] that came up when applied to real functions: await in nested expr; complex lifetimes.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

[ci no changelog needed] no user-visible changes

### CODE REVIEW

**Comment by kazcw** [comment]
Hm. The compiler segfaults when building the updated enso-profiler tests:

`0x00007ffff6427db9 in _RNvYNtNtNtCsDQ5ZK7eIcO_16rustc_ast_pretty6pprust5state5StateNtB4_10PrintState9print_ttsB8_.llvm.18266741937534307445 () from /usr/lib/librustc_driver-381486f3b973966d.so`

Seems like a rustc bug. Will be looking into this tomorrow morning.

**Comment by kazcw** [comment]
Some real data, as represented by the measurements pretty-printer:
- `entry_point_ide` and its descendants measure time to load the project-chooser scene. It's <150ms long, but starts after 3 seconds of loading (mostly compiling WASM I think).
- `create_new_project` represents the results of clicking on a project template. It takes 5.3s for it and its descendants to finish, except `notify_about_compiling_process`, which finishes an additional 3.5s later.
- The top-level `setup_and_display_new_project` should be a child of `create_new_project`; I have achieved this in a sub-branch by attaching profilers to the notification objects, but I'm going to PR that separately because it relies on some changes to the low-level API.

The timing breakdown will be more visible in the graph, but this shows the hierarchy better.

```
entry_point_ide (app/gui/src/lib.rs:121)
  interval: 3096.8-3097.8
  children:
    start_and_forget (app/gui/src/ide/initializer.rs:62)
      interval: 3097.6-3097.8
      children:
        start_and_forget (app/gui/src/ide/initializer.rs:65)
          intervals:
            interval: 3111.6-3111.6
            interval: 3195.4-3195.4
          children:
            start (app/gui/src/ide/initializer.rs:78)
              intervals:
                interval: 3111.6-3136.7
                interval: 3195.2-3195.4
              children:
                initialize_ide_controller (app/gui/src/ide/initializer.rs:142)
                  intervals:
                    interval: 3136.7-3136.7
                    interval: 3195.2-3195.2
                    interval: 3195.2-3195.2
                  children:
                    setup_project_manager (app/gui/src/ide/initializer.rs:175)
                      intervals:
                        interval: 3136.7-3136.7
                        interval: 3195.2-3195.2
                      children:
                new (app/gui/src/presenter.rs:145)
                  interval: 3195.3-3195.3
                  children:
                    init (app/gui/src/presenter.rs:165)
                      interval: 3195.3-3195.3
                      children:
                        set_projects_list_on_welcome_screen (app/gui/src/presenter.rs:218)
                          intervals:
                            interval: 3195.4-3195.4
                            interval: 3237.9-3238.7
                          children:
                            list_projects (app/gui/src/controller/ide/desktop.rs:160)
                              intervals:
                                interval: 3195.4-3195.5
                                interval: 3237.9-3237.9
                              children:
                        setup_and_display_new_project (app/gui/src/presenter.rs:40)
                          interval: 3195.3-3195.3
                          children:
create_new_project (app/gui/src/controller/ide/desktop.rs:136)
  intervals:
    interval: 6299.2-6299.2
    interval: 6353.5-6353.7
    interval: 7311.9-7311.9
    interval: 9995.0-9995.0
  children:
    new_opened (app/gui/src/model/project/synchronized.rs:342)
      intervals:
        interval: 7311.9-7311.9
        interval: 8745.1-8745.1
        interval: 9995.0-9995.0
      children:
        new_connected (app/gui/src/model/project/synchronized.rs:307)
          intervals:
            interval: 8745.1-8745.3
            interval: 8762.1-8762.1
            interval: 8778.6-8778.7
            interval: 9795.1-9795.1
            interval: 9811.8-9811.8
            interval: 9995.0-9995.0
          children:
            new (app/gui/src/model/project/synchronized.rs:251)
              intervals:
                interval: 9811.8-9811.8
                interval: 9978.4-9978.5
                interval: 9995.0-9995.0
              children:
setup_and_display_new_project (app/gui/src/presenter.rs:40)
  interval: 9995.0-10013.1
  children:
    setup_and_display (app/gui/src/presenter.rs:54)
      intervals:
        interval: 10013.1-10013.3
        interval: 11668.6-11668.6
      children:
        initialize (app/gui/src/presenter/project.rs:279)
          intervals:
            interval: 10013.3-10013.3
            interval: 10928.7-11668.6
          children:
            initialize (app/gui/src/controller/project.rs:123)
              intervals:
                interval: 10013.3-10013.3
                interval: 10245.2-10245.2
                interval: 10835.0-10835.0
                interval: 10835.0-10835.0
                interval: 10928.7-10928.7
                interval: 10928.7-10928.7
              children:
                notify_about_compiling_process (app/gui/src/controller/project.rs:218)
                  intervals:
                    interval: 11668.6-11668.6
                    interval: 15179.4-15179.4
                  children:
            new (app/gui/src/presenter/project.rs:169)
              interval: 10928.7-11668.6
              children:
                new (app/gui/src/presenter/project.rs:37)
                  interval: 10928.7-11668.6
                  children:
                    new (app/gui/src/presenter/graph.rs:455)
                      interval: 10928.7-11668.6
                      children:
                        init (app/gui/src/presenter/graph.rs:467)
                          interval: 10928.8-11668.6
                          children:
                    new (app/gui/src/presenter/code.rs:73)
                      interval: 11668.6-11668.6
                      children:
                init (app/gui/src/presenter/project.rs:182)
                  interval: 11668.6-11668.6
                  children:
```

File: app/gui/src/controller/ide/desktop.rs
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs]: This is very confusing - we are defining pairs of functions with the same name here. We can call some of them `..._internal` or anything, but for just clarity, the names should be easily distinguishable.

File: app/gui/src/controller/ide/desktop.rs
Comment (wdanilo) [inline app/gui/src/controller/ide/desktop.rs:136]: these profilers are freaking awesome

File: app/gui/src/ide/initializer.rs
Comment (wdanilo) [inline app/gui/src/ide/initializer.rs]: This is confusing IMO too. We are defining a local function of the same name as the parent function. Are we able to change one of the names? Let's find a rule how to rename such functions and lets follow this rule. This is way less confusing than the previous one though (previous = defining overlapping methods on structs), so if there is no better idea, we can leave it.

File: app/gui/src/lib.rs
Comment (wdanilo) [inline app/gui/src/lib.rs]: What do you think about replacing it with `pub use enso_profiler::prelude::*;`? We are gathering this way "popular exports from libraries" and this might be a little bit more handy in other libs that would like to include profiler in their local preludes.

File: build/run.js
Comment (wdanilo) [inline build/run.js]: Could you explain it better, please? In particular:
1. What does this feature do - shortly for reader reading the code (I assume it leaves line numbers in debug symbols in the resulting binary?)
2. What do you mean by "breaks JetBrains IDEs"? What exactly is happening? Please link an issue from their bug tracker.
3. What do you mean by "it is disabled in the default features" – what are "default features" and where are they defined?
4. Why enabling it here works, while when its enabled in other place it does not? I feel I'm missing something super simple here.

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: not needed anymore

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: This should have documentation explaining what it does.

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: This does not make sense. `add_child` removes the child from the previous parent. Moreover, the display object of world is the same as the default scene, so this is duplicated line.

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: why are you cloning camera, only to pass it as a reference in the next line?

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: This is not needed. World owns scene, and you already mom-forgot it.

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: refactor it to variables, please, so it will become 1-line expressions.

File: lib/rust/ensogl/example/render-profile/src/lib.rs
Comment (wdanilo) [inline lib/rust/ensogl/example/render-profile/src/lib.rs]: duplicate literal

File: lib/rust/profiler/data/src/bin/measurements.rs
Comment (wdanilo) [inline lib/rust/profiler/data/src/bin/measurements.rs:24]: no docs in the header. It means that linter is not enabled here. Please check the linter configuration.

File: lib/rust/profiler/data/src/bin/measurements.rs
Comment (wdanilo) [inline lib/rust/profiler/data/src/bin/measurements.rs:28]: This should have docs

File: lib/rust/profiler/data/src/bin/measurements.rs
Comment (wdanilo) [inline lib/rust/profiler/data/src/bin/measurements.rs:36]: This should have docs

File: lib/rust/profiler/macros/build.rs
Comment (wdanilo) [inline lib/rust/profiler/macros/build.rs:15]: No module docs. Please enable linter here.

File: lib/rust/profiler/macros/src/lib.rs
Comment (wdanilo) [inline lib/rust/profiler/macros/src/lib.rs:423]: Thank you for providing comments with clear explanation in the code! It's amazing!

**Review by wdanilo — APPROVED** [review]

File: app/gui/src/ide/initializer.rs
Comment (kazcw) [reply app/gui/src/ide/initializer.rs]: As I'm instrumenting more details this is coming up a lot, so I have implemented a better solution for the upcoming PR: I'm extending `#[profile]` to create child profilers for any async blocks it encounters.

For now I will use "_internal" names to get this merged and usable.

File: build/run.js
Comment (kazcw) [reply build/run.js]: Updated with more explanation. For context, this is part of the solution to the problems I had with the initial version of `#[profiler]` interfering with IntelliJ's understanding of the code.

**Comment by kazcw** [comment]
This needs review from a Cargo.lock codeowner, because it involves creating a new crate. @MichaelMauderer?

@wdanilo Shall I add you to CODEOWNERS for Cargo.lock?

**Comment by kazcw** [comment]
@MichaelMauderer @4e6 @mwu-tow @farmaazon Could someone approve my changes to Cargo.lock (added a new crate)? @wdanilo is not a codeowner of that file.

**Review by mwu-tow — APPROVED** [review]
Accepted wrt Cargo.lock.

**Comment by kazcw** [comment]
@mwu-tow a CI problem is blocking merge here. The Windows build keeps failing with "SignTool Error: No certificates were found that met all the given criteria." I thought it could be a transient failure but I've tried rerunning CI several times.

**Comment by mwu-tow** [comment]
@kazcw I am forcing merge. This is a quite old issue with signing, so far it usually went away after a retry. Apparently, over the time it got worse.

**Pull request merged**
